src/test/java/org/xbill/DNS/LookupTest.java

```diff
@@ -2,89 +2,345 @@
+    wireUpMockResolver(mockResolver, this::simpleAnswer);
+
+    Lookup lookup = makeLookupWithResolver(mockResolver, "host.subdomain");
+    lookup.setSearchPath("example.com");
+    Record[] results = lookup.run();
+
+    ArgumentCaptor<Message> messageCaptor = ArgumentCaptor.forClass(Message.class);
+    verify(mockResolver).send(messageCaptor.capture());
+
+    assertEquals(
+        Record.newRecord(Name.fromConstantString("host.subdomain."), Type.A, DClass.IN, 0L),
+        messageCaptor.getValue().getSection(Section.QUESTION).get(0));
+
+    assertEquals(1, results.length);
+  }
+
+  @Test
+  void testRun_firstSearchPathAppendedHitsCNAME() throws Exception {
+    wireUpMockResolver(mockResolver, this::maybeCnameAnswer);
+
+    Lookup lookup = makeLookupWithResolver(mockResolver, "host");
+    lookup.setSearchPath("first.example.com", "second.example.com");
+    Record[] results = lookup.run();
+
+    ArgumentCaptor<Message> messageCaptor = ArgumentCaptor.forClass(Message.class);
+    verify(mockResolver, times(2)).send(messageCaptor.capture());
+
+    List<Message> queries = messageCaptor.getAllValues();
+
+    assertEquals(
+        Record.newRecord(Name.fromConstantString("host.first.example.com."), Type.A, DClass.IN, 0L),
+        queries.get(0).getQuestion());
+    assertEquals(
+        Record.newRecord(Name.fromConstantString("target.example.com."), Type.A, DClass.IN, 0L),
+        queries.get(1).getQuestion());
+
+    assertEquals(1, results.length);
+  }
+
+  Message maybeCnameAnswer(Message query) {
+    return answer(
+        query,
+        name -> {
+          if (name.toString().equals("host.first.example.com.")) {
+            return new CNAMERecord(
+                DUMMY_NAME, DClass.IN, 0, Name.fromConstantString("target.example.com."));
+          }
+          return new ARecord(DUMMY_NAME, DClass.IN, 0, InetAddress.getLoopbackAddress());
+        });
+  }
+
+  @Test
+  void testRun_CNAMELoop() throws Exception {
+    wireUpMockResolver(mockResolver, this::cnameLoopAnswer);
+
+    Lookup lookup = makeLookupWithResolver(mockResolver, "host");
+    lookup.setSearchPath("first.example.com", "second.example.com");
+    Record[] results = lookup.run();
+
+    assertNull(results);
+    assertEquals(Lookup.UNRECOVERABLE, lookup.getResult());
+    assertEquals("CNAME loop", lookup.getErrorString());
+  }
+
+  Message cnameLoopAnswer(Message query) {
+    return answer(
+        query,
+        name -> {
+          if (name.toString().equals("first.example.com.")) {
+            return new CNAMERecord(
+                DUMMY_NAME, DClass.IN, 0, Name.fromConstantString("second.example.com."));
+          }
+          return new CNAMERecord(
+              DUMMY_NAME, DClass.IN, 0, Name.fromConstantString("first.example.com."));
+        });
+  }
+
+  @Test
+  void testRun_reuseLookup() throws Exception {
+    wireUpMockResolver(mockResolver, this::simpleAnswer);
+
+    Lookup lookup = makeLookupWithResolver(mockResolver, "host");
```

Comment:
Same here, ndots at 1 means relative lookup. There might be more, I didn't thoroughly look at all tests.